in_memory_document_session_operations.go

```diff
@@ -0,0 +1,847 @@
+package ravendb
+
+import (
+	"fmt"
+	"reflect"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	// TODO: fix names
+	MetadataCOLLECTION    = "@collection"
+	MetadataPROJECTION    = "@projection"
+	MetadataKEY           = "@metadata"
+	MetadataID            = "@id"
+	MetadataCONFLICT      = "@conflict"
+	MetadataID_PROPERTY   = "Id"
+	MetadataFLAGS         = "@flags"
+	MetadataATTACHMENTS   = "@attachments"
+	MetadataINDEX_SCORE   = "@index-score"
+	MetadataLAST_MODIFIED = "@last-modified"
+	MetadataRAVEN_GO_TYPE = "Raven-Go-Type"
+	MetadataCHANGE_VECTOR = "@change-vector"
+	MetadataEXPIRES       = "@expires"
+)
+
+// SessionInfo describes a session
+type SessionInfo struct {
+	SessionID int
+}
+
+var (
+	_clientSessionIDCounter int32 = 1
+)
+
+func newClientSessionID() int {
+	newID := atomic.AddInt32(&_clientSessionIDCounter, 1)
+	return int(newID)
+}
+
+// InMemoryDocumentSessionOperations represents database operations queued
+// in memory
+type InMemoryDocumentSessionOperations struct {
+	_clientSessionID int
+	deletedEntities  map[interface{}]struct{}
+	RequestExecutor  *RequestExecutor
+	// TODO: OperationExecutor
+	// Note: pendingLazyOperations and onEvaluateLazy not used
+	generateDocumentKeysOnStore bool
+	sessionInfo                 SessionInfo
+	_saveChangesOptions         *BatchOptions
+
+	// Note: skipping unused isDisposed
+	id string
+
+	/* TODO:
+	   private final List<EventHandler<BeforeStoreEventArgs>> onBeforeStore = new ArrayList<>();
+	   private final List<EventHandler<AfterSaveChangesEventArgs>> onAfterSaveChanges = new ArrayList<>();
+	   private final List<EventHandler<BeforeDeleteEventArgs>> onBeforeDelete = new ArrayList<>();
+	   private final List<EventHandler<BeforeQueryEventArgs>> onBeforeQuery = new ArrayList<>();
+	*/
+
+	// ids of entities that were deleted
+	_knownMissingIds map[string]struct{}
+
+	// Note: skipping unused externalState
```

Comment:
We want to provide this so users can attach state for the session if they need to.
This can make it easy to support event handling, for example.

Reply:
It isn't commonly used, however, so can be lazily defined.